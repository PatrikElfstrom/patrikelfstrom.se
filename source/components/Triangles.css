.triangles {
  height: 100%;
  width: 100%;
  min-width: 100%;
  min-height: 500px;
  position: absolute;
  top: 0;
  left: 0;
  z-index: -1;
  overflow: hidden;
  background: #212121;
}

.triangles svg {
  position: absolute;
  height: 100%;
  width: 100%;
  left: 0;
  top: 0;
}

.triangles path {
  transition: all 500ms 1000ms ease-in-out;
}
