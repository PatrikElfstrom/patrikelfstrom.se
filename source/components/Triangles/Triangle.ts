import { Polygon, Point } from '@pixi/math';
import { settings, SmoothGraphics as Graphics } from '@pixi/graphics-smooth';
import { Renderer, RenderTexture } from '@pixi/core';
import { Sprite as PixiSprite } from '@pixi/sprite';
import type { Sprite, TriangleSize } from '../../types';

settings.PIXEL_LINE = 1;

export class Triangle {
  polygon!: Polygon;

  graphics!: Graphics;

  texture!: RenderTexture;

  sprite!: Sprite;

  width!: number;

  height!: number;

  constructor(size: number) {
    const { width, height } = Triangle.calculateSize(size);
    this.polygon = Triangle.createPolygon({ width, height });

    this.width = width;
    this.height = height;
  }

  static calculateSize(width: number): TriangleSize {
    const height = Math.round(Math.sqrt(width ** 2 - (width / 2) ** 2));

    return { width, height };
  }

  static createPolygon({ width, height }: TriangleSize): Polygon {
    return new Polygon([new Point(0, height), new Point(width / 2, 0), new Point(width, height)]);
  }

  generateGraphics(color: number): Graphics {
    const graphics = new Graphics();

    // Defringe edges
    graphics.lineStyle({
      color,
      width: 1,
      alignment: 0.5,
      alpha: 1,
    });

    graphics.beginFill(color, 1, true);
    graphics.drawShape(this.polygon);
    graphics.endFill();

    return graphics;
  }

  static generateTexture(graphics: Graphics, renderer: Renderer): RenderTexture {
    return renderer.generateTexture(graphics);
  }

  createSprite(texture: RenderTexture): Sprite {
    const sprite = PixiSprite.from(texture) as Sprite;

    // Rotate 90 deg
    sprite.rotation = Math.PI / 2;

    // Set sprite anchor to middle
    sprite.anchor.set(0.5);

    // Reset position because of change to anchor
    sprite.x += this.height / 2;
    sprite.y += this.width / 2;

    // Since we changed the anchor
    // We need to recalculate the hit area
    sprite.hitArea = this.calculateSpriteHitArea();

    // Hide as default
    sprite.alpha = 0;

    // Set default age
    sprite.age = 0;

    return sprite;
  }

  calculateSpriteHitArea(): Polygon {
    return new Polygon([
      new Point(-this.width / 2, this.height / 2),
      new Point(0, -this.height / 2),
      new Point(this.width / 2, this.height / 2),
    ]);
  }
}
