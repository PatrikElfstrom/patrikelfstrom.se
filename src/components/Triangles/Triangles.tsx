import { ReactElement, useEffect, useRef } from 'react';
import { TriangleGrid } from './triangleGrid';
import './style.css';

export default function Triangles(): ReactElement {
  const canvas = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    if (canvas.current) {
      new TriangleGrid({
        backgroundColor: 0x21_21_21,
        resizeTo: window.document.body,
        size: 40,
        view: canvas.current,
        resolution: window.devicePixelRatio || 1,
        autoDensity: true,
      });
    }
  }, [canvas]);

  return <canvas className="canvas" ref={canvas} />;
}
