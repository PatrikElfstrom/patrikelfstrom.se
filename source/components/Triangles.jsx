import React, { useRef, useEffect } from 'react';
import Triangles from '../lib/Triangles';

const TriangleBackground = () => {
  const canvas = useRef(null);

  useEffect(() => {
    const triangles = new Triangles({
      backgroundColor: 0x212121,
      resizeTo: window,
      size: 40,
      view: canvas.current,
    });
  }, []);

  return <canvas ref={canvas}>asd</canvas>;
};

export default TriangleBackground;
