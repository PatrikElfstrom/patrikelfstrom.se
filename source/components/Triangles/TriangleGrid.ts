import { Renderer, BatchRenderer } from '@pixi/core';
import { Application } from '@pixi/app';
import { Container } from '@pixi/display';
import { EventEmitter } from '@pixi/utils';
import { TickerPlugin } from '@pixi/ticker';
import { debounce } from 'lodash-es';
import { randomHslGenerator, hslToHex } from '../../helpers/colors';
import { Triangle } from './Triangle';
import { Textures, TriangleRenderCallback, Positions, TrianglesOptions, Sprite } from '../../types';
import { randomNumber } from '../../helpers/numbers';

Renderer.registerPlugin('batch', BatchRenderer);
Application.registerPlugin(TickerPlugin);

export class TriangleGrid extends Application {
  trianglePositions: Positions[] = [];

  logoTrianglePositions: Positions[] = [];

  triangles: Promise<Sprite>[] = [];

  triangle: Triangle;

  size!: number;

  trianglesPerRow!: number;

  trianglesPerColumn!: number;

  container: Container;

  textures: Textures = {};

  events: EventEmitter;

  renderQueue: boolean[] = [];

  centerPosition!: Positions;

  constructor(options: TrianglesOptions) {
    super(options);

    this.ticker.stop();

    this.events = new EventEmitter();

    this.triangle = new Triangle(options.size);

    // Create a container
    this.container = new Container();

    // Add container to stage
    this.stage.addChild(this.container);

    this.generateTriangles();

    window.addEventListener('resize', debounce(this.generateTriangles, 100));
  }

  generateTriangles = (): void => {
    this.renderQueue.push(true);
    this.ticker.start();

    // Remove existing children so we don't get duplicates
    this.container.removeChildren();

    ({ trianglesPerRow: this.trianglesPerRow, trianglesPerColumn: this.trianglesPerColumn } =
      this.calculateNumberOfTriangles());

    let isLeftPointing = false;
    let offsets = 0;
    while (!isLeftPointing) {
      this.trianglePositions = this.calculateTrianglePositions();
      this.centerPosition = this.getCenterPosition(
        this.trianglesPerRow,
        this.trianglesPerColumn,
        offsets,
      );

      isLeftPointing = this.centerPosition?.scale.y === -1;
      offsets += 1;
    }

    this.logoTrianglePositions = this.calculateLogoTrianglePositions();

    this.triangles = this.renderTriangles(({ color, positions, delay }) => {
      if (
        this.logoTrianglePositions.some(
          (trianglePosition) =>
            trianglePosition?.grid?.x === positions?.grid?.x &&
            trianglePosition?.grid?.y === positions?.grid?.y,
        )
      ) {
        delay = randomNumber(500, 2000);
        color = hslToHex(randomHslGenerator(200, 100, 30, 0, 10));
      }

      return { color, positions, delay };
    });

    // Stop ticker when all triangles are done
    Promise.all(this.triangles)
      .then(() => this.renderQueue.pop())
      // stop ticker if there is nothing rendering
      .then(() => this.renderQueue.length <= 0 && this.ticker.stop())
      .catch((error) => {
        throw error;
      });
  };

  calculateNumberOfTriangles(): {
    trianglesPerRow: number;
    trianglesPerColumn: number;
  } {
    const initialTrianglesPerRow = this.screen.width / this.triangle.height;

    // Round to nearest even number
    const trianglesPerRow = Math.round(initialTrianglesPerRow / 2) * 2;

    const triangleHeight = this.screen.width / trianglesPerRow;

    // Set new triangle height
    this.triangle.height = triangleHeight;

    // Divide by half since each triangle interlock
    // add one since we see the next row
    const trianglesPerColumn = Math.ceil(this.screen.height / (this.triangle.width / 2)) + 1;

    return { trianglesPerRow, trianglesPerColumn };
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  getCenterPosition(trianglesPerRow: number, trianglesPerColumn: number, offset: number) {
    // Calculate the center position
    const positionX = Math.round(trianglesPerRow / 2);
    let positionY = Math.round(trianglesPerColumn / 2);

    // Move the center position up a bit
    const goldenRatio = (1 + Math.sqrt(5)) / 2;
    positionY = Math.ceil(positionY / goldenRatio) - offset;

    const centerPosition = this.trianglePositions.find(
      (position) => position?.grid?.y >= positionY && position?.grid?.x === positionX,
    );

    if (!centerPosition) {
      throw new Error('Could not find center position');
    }

    return centerPosition;
  }

  calculateLogoTrianglePositions(): Positions[] {
    if (!this.centerPosition) {
      throw new Error('No center position');
    }

    const positionX = this.centerPosition.grid.x;
    const positionY = this.centerPosition.grid.y;

    // The logo design
    const triangles = [
      // Center
      { y: 0, x: 0 },
      { y: -1, x: 0 },
      { y: -1, x: -1 },
      { y: 0, x: -1 },
      { y: 1, x: -1 },
      { y: 1, x: 0 },

      // Ring Left
      { y: -5, x: -1 },
      { y: -4, x: -1 },
      { y: -4, x: -2 },
      { y: -3, x: -2 },
      { y: -3, x: -3 },
      { y: -2, x: -3 },
      { y: -1, x: -3 },
      { y: 0, x: -3 },
      { y: 1, x: -3 },
      { y: 2, x: -3 },
      { y: 3, x: -3 },

      // Ring right
      { y: -5, x: 0 },
      { y: -4, x: 0 },
      { y: -4, x: 1 },
      { y: -3, x: 1 },
      { y: -3, x: 2 },
      { y: -2, x: 2 },
      { y: -1, x: 2 },
      { y: 0, x: 2 },
      { y: 1, x: 2 },
      { y: 2, x: 2 },
      { y: 3, x: 2 },
      { y: 3, x: 1 },
      { y: 4, x: 1 },
      { y: 4, x: 0 },
      { y: 5, x: 0 },
      { y: 5, x: -1 },
      { y: 4, x: -1 },
    ];

    return triangles.map((triangle) => {
      const xIndex = positionX + triangle.x;
      const yIndex = positionY + triangle.y;
      let xPixel = xIndex * this.triangle.height;
      const yPixel = (yIndex * this.triangle.width) / 2;

      // Add 50% because of reasons
      xPixel += this.triangle.height / 2;

      // Flip every even column in every odd row
      // And flip every odd column in every even row
      const yScale =
        (yIndex % 2 === 0 && xIndex % 2 === 1) || (yIndex % 2 === 1 && xIndex % 2 === 0) ? -1 : 1;

      return {
        grid: { x: xIndex, y: yIndex },
        pixel: {
          x: xPixel,
          y: yPixel,
        },
        scale: { x: 1, y: yScale },
      };
    });
  }

  calculateTrianglePositions(): Positions[] {
    const trianglePositions = [];

    for (let xIndex = 0; xIndex < this.trianglesPerRow; xIndex += 1) {
      for (let yIndex = 0; yIndex < this.trianglesPerColumn; yIndex += 1) {
        const position: Positions = {
          grid: { x: xIndex, y: yIndex },
          pixel: { x: 0, y: 0 },
          scale: { x: 1, y: 1 },
        };

        // Flip every even column in every odd row
        if (yIndex % 2 === 0 && xIndex % 2 === 1) {
          position.scale.y = -1;
        }

        // Flip every odd column in every even row
        if (yIndex % 2 === 1 && xIndex % 2 === 0) {
          position.scale.y = -1;
        }

        // Move the each row 50% to the right
        // Since we start from the triangle center
        position.pixel.x += this.triangle.height / 2;

        // Position triangle
        position.pixel.x += xIndex * this.triangle.height;
        position.pixel.y += (yIndex * this.triangle.width) / 2;

        trianglePositions.push(position);
      }
    }

    return trianglePositions;
  }

  renderTriangles(triangleRenderCallback?: TriangleRenderCallback): Promise<Sprite>[] {
    return this.trianglePositions.map(
      (trianglePosition) =>
        new Promise((resolve) => {
          const hexColor = hslToHex(randomHslGenerator());
          const visibilityDelay = randomNumber(0, 500);

          const { color, positions, delay } =
            typeof triangleRenderCallback === 'function'
              ? triangleRenderCallback({
                  color: hexColor,
                  positions: trianglePosition,
                  delay: visibilityDelay,
                })
              : {
                  color: hexColor,
                  positions: trianglePosition,
                  delay: visibilityDelay,
                };

          // Only generate if no texture with the same color has been generated
          if (!Object.prototype.hasOwnProperty.call(this.textures, color)) {
            const graphics = this.triangle.generateGraphics(color);
            const texture = Triangle.generateTexture(graphics, this.renderer as Renderer);

            this.textures[color] = texture;
          }

          const sprite = this.triangle.createSprite(this.textures[color]);

          sprite.scale.y = positions.scale.y;
          sprite.scale.x = positions.scale.x;
          sprite.x = positions.pixel.x;
          sprite.y = positions.pixel.y;

          const spriteTargetAlpha = 1;

          const transitionSpeed = 1; // In seconds
          const fps = 60;
          const transitionSpeedPerFrame = (fps * transitionSpeed) / 1000;

          const showSprite = () => {
            sprite.age = Date.now() - sprite.added;

            // Delay sprite
            if (sprite.age >= delay) {
              // Remove ticker if sprite is fully visible
              if (sprite.alpha < spriteTargetAlpha) {
                sprite.alpha += transitionSpeedPerFrame * this.ticker.deltaTime;
              } else {
                this.ticker.remove(showSprite);
                resolve(sprite);
              }
            }
          };

          this.ticker.add(showSprite);

          sprite.on('added', () => {
            sprite.added = Date.now();
          });

          this.container.addChild(sprite);
        }),
    );
  }
}
