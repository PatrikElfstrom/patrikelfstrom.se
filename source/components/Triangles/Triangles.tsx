import { ReactElement, useEffect, useRef } from 'react';
import { TriangleGrid } from './TriangleGrid';

export const Triangles = (): ReactElement => {
  const canvas = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    if (canvas.current) {
      new TriangleGrid({
        backgroundColor: 0x21_21_21,
        resizeTo: window.document.body,
        size: 40,
        view: canvas.current,
      });
    }
  }, []);

  return <canvas ref={canvas} />;
};
