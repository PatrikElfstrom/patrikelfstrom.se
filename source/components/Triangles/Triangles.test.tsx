import TestRenderer from 'react-test-renderer';
import { TriangleGrid } from './TriangleGrid';
import { Triangles } from './Triangles';

describe('<Triangles />', () => {
  it('should match snapshot', () => {
    const tree = TestRenderer.create(<Triangles />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
