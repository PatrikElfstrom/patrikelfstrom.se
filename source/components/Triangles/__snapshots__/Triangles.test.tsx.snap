// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Triangles /> should match snapshot 1`] = `<canvas />`;
